graph-builder/src/graph.rs

```diff
@@ -306,49 +306,55 @@ pub fn create_graph(releases: Vec<registry::Release>) -> Result<Graph, Error> {
     releases
         .into_iter()
         .inspect(|release| trace!("Adding a release to the graph '{:?}'", release))
-        .try_for_each(|release| {
-            let previous = release
-                .metadata
-                .previous
-                .iter()
-                .cloned()
+        .map(|release| {
+            Ok((
+                release.metadata.next.clone(),
+                release.metadata.previous.clone(),
+                release.metadata.version.build.clone(),
+                graph.add_release(release)?,
+            ))
+        })
+        .collect::<Vec<Fallible<_>>>()
+        .into_iter()
+        .try_for_each(|result| {
+            let (next, previous, current_build, current) = result?;
+
+            previous
+                .into_iter()
                 .map(|mut previous| {
-                    previous.build = release.metadata.version.build.clone();
+                    previous.build = current_build.clone();
                     previous
                 })
-                .collect::<Vec<_>>();
-
-            let next = release
-                .metadata
-                .next
-                .iter()
-                .cloned()
+                .try_for_each(|version| {
+                    let previous = match graph.find_by_version(&version.to_string()) {
+                        Some(id) => id,
+                        None => {
+                            warn!("Adding abstract release for {}", version.to_string());
+                            graph.add_release(Release::Abstract(AbstractRelease {
```

Comment:
if you add all the releases first, do you still need the concept of `AbstractRelease`?

Reply:
@crawford thinks they could be useful if an edge refers to a release which was never officially released, but might have made it to specific clusters.